src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,351 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use std::sync::Arc;
+use tokio::sync::{Mutex, Notify};
+// Using tokio's wrapper makes the cache testable without relying on clock times.
+use tokio::time::Instant;
+
+// TODO(#1210) - implementation in progress
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider,
+{
+    // The cached token, or the last seen error.
+    token: Arc<Mutex<Result<Token>>>,
+
+    // Tracks if a refresh is ongoing. If the lock is held, there is a refresh.
+    refresh_in_progress: Arc<Mutex<()>>,
+    // Allows us to await the result of a refresh in multiple threads.
+    refresh_notify: Arc<Notify>,
+
+    // The token provider. This thing does the refreshing.
+    inner: Arc<T>,
+}
+
+// Returns true if we are holding an error, or a token that has expired.
+fn invalid(token: &Result<Token>) -> bool {
+    match token {
+        Ok(t) => t.expires_at.is_some_and(|e| e <= Instant::now().into_std()),
+        Err(_) => true,
+    }
+}
+
+// We manually implement the `Clone` trait because the Rust compiler will
+// squawk if `T` is not `Clone`, even though we only hold an `Arc<T>`. :shrug:
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            token: self.token.clone(),
+            refresh_in_progress: self.refresh_in_progress.clone(),
+            refresh_notify: self.refresh_notify.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T: TokenProvider> TokenCache<T> {
+    // TODO(#1210) - implementation in progress
+    #[allow(dead_code)]
+    pub fn new(inner: T) -> TokenCache<T> {
+        TokenCache {
+            token: Arc::new(Mutex::new(Err(crate::errors::CredentialError::retryable_from_str("No token in the cache. This should never happen. Something has gone wrong. Open an issue at https://github.com/googleapis/google-cloud-rust/issues/new?template=bug_report.md.")))),
+            refresh_in_progress: Arc::new(Mutex::new(())),
+            refresh_notify: Arc::new(Notify::new()),
+            inner: Arc::new(inner),
+        }
+    }
+
+    // Clones the current token, in a thread-safe manner. Releases the lock on return.
+    async fn current_token(&self) -> Result<Token> {
+        self.token.lock().await.clone()
```

Comment:
Does rust have concept of read lock vs write lock? We would need to lock only when the token is getting updated I believe?

Reply:
Yes, good catch! Fixed.

I had `RwLock` initially, but changed it when I was flailing about trying to fix my unit tests. We do not need `tokio::sync::Mutex` for this because we do not need to hold the lock over an `await` boundary.

I also added another unit test to exercise the read lock contention.